[AMDGPU] don't mark control-flow intrinsics as convergent

ssahasra opened the pull request:
This is really a workaround to allow control flow lowering in the presence of convergence control tokens. Control-flow intrinsics in LLVM IR are convergent because they indirectly represent the wave CFG, i.e., sets of threads that are "converged" or "execute in lock-step". But they exist during a small window in the lowering process, inserted after the structurizer and then translated to equivalent MIR pseudos. So rather than create convergence tokens for these builtins, we simply mark them as not convergent.

The corresponding MIR pseudos are marked as having side effects, which is sufficient to prevent optimizations without having to mark them as convergent.

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -3143,25 +3143,37 @@ def int_amdgcn_cvt_sr_fp8_f32 : ClangBuiltin<"__builtin_amdgcn_cvt_sr_fp8_f32">,
 // should emit calls to these.
 // ===----------------------------------------------------------------------===//
+//
+// This is really a workaround to allow control flow lowering in the presence of
```
This comment starts with no context

ssahasra replied:
Fixed.

ssahasra committed:
[AMDGPU] don't mark control-flow intrinsics as convergent

This is really a workaround to allow control flow lowering in the presence of
convergence control tokens. Control-flow intrinsics in LLVM IR are convergent
because they indirectly represent the wave CFG, i.e., sets of threads that are
"converged" or "execute in lock-step". But they exist during a small window in
the lowering process, inserted after the structurizer and then translated to
equivalent MIR pseudos. So rather than create convergence tokens for these
builtins, we simply mark them as not convergent.

The corresponding MIR pseudos are marked as having side effects, which is
sufficient to prevent optimizations without having to mark them as convergent.

ssahasra commented:
Bump! Added @alex-t because I am told he is working on changes to the control flow lowering.
